1-Draft/RFCNNNN-Download-Cmdlet.md

```diff
@@ -0,0 +1,294 @@
+* Current Directory: `c:\images\`
+* Directory files: none
+* Current Time: 2018-03-31 08:11:29.632
+* File name: `c:\images\gXnRf4J.jpg`
+
+Example 6:
+
+* Uri: `https://i.imgur.com/gXnRf4J>.jpg`
+* Current Directory: `c:\images\`
+* Directory files: none
+* Current Time: 2018-03-31 08:11:29.632
+* File name: `c:\images\gXnRf4J%3e.jpg`
+
+### PassThru Feature
+
+By default the cmdlet will return no output.
+The cmdlet is specialized for downloading.
+The cmdlet will perform no content encoding or decoding.
+The cmdlet only works with raw bytes.
+
+As the cmdlets primary function is to create a local file from a remote file,
+The only output it shall provide is the file it creates.
+
+By issuing the `-PassThru` switch,
+the cmdlet will return the equivalent of `Get-Item`.
+This will allow the cmdlet to pass the item to commands such as `Get-Content`
+where the user can specify what encoding features, if any, they require.
+
+This allows the cmdlet code to focus only on download features and optimizations
+without being coupled to the output ecosystem.
+For example, should the language adopt UTF-32 as the standard encoding,
+the cmdlet will not be coupled to that change.
+
+### AsJob Feature
+
+By default the cmdlet will run synchronously and block the current thread until completion.
+By supplying `-AsJob`, the user can run the cmdlet as a background job.
+As file downloads may take significant time,
+the ability to begin the download and perform other tasks should be native to the cmdlet.
+
+When supplied, the cmdlet will output a Job object.
+This feature will work in a manner similar to other cmdlets which implement it.
+
+### Multiple Protocol Support
+
+The cmdlet shall support multiple protocols.
+The cmdlet shall be extensible to support additional protocols.
+The protocol handlers shall be extensible to handle additional features.
+
+The extensibility is accomplished by interfaces.
+
+```csharp
+internal interface IInvokeDownloadProtocol
+{
+    Stream GetStream(PSCmdlet Cmdlet);
+    Uri GetUri();
+    void SetUri(Uri RequestUri);
+}
+
+internal interface IInvokeDownloadResumable
+{
+    void SetResume(bool Resume);
+    bool GetResume();
+    bool IsResumeSuccessful();
+    bool IsResumeComplete();
+}
+
+internal interface IInvokeDownloadAuthenticator
+{
+    void SetCredential(PSCredential Credential);
+    PSCredential GetCredential();
+}
+
+internal interface IInvokeDownloadProxyable
+{
+    void SetProxy(Uri ProxyUri);
+    Uri GetProxy();
+    void SetProxyCredential(PSCredential Credential);
+    PSCredential GetProxyCredential();
+}
+
+internal interface IInvokeDownloadRedirectable
+{
+    IList<Uri> GetRedirectionChain(PSCmdlet Cmdlet, Uri RequestUri);
+}
+```
+
+Class implementing these interfaces will be registered to a static internal dictionary
+
+```csharp
+public class InvokeDownloadCmdlet : PSCmdlet
+{
+    internal static Dictionary<String, IInvokeDownloadProtocol> ProtocolHandlers;
+}
+```
```

Comment:
Both external and internal implementations of those interfaces need to somehow register to the cmdlet. How would the registration work?
- You could expose a static methods from the cmdlet type for a module to call to register its implementation types. Then the module needs to implement `IModuleAssemblyInitializer.OnImport` to register the protocol when it's being imported, and maybe unregister via `IModuleAssemblyCleanup.OnRemove` when it's being removed?
- Add another cmdlet for registration? Maybe not really needed ...

What are your thoughts?

Reply:
The way how `JobManager` deals with 3rd party `JobSourceAdapter` types is to [register them when importing the module](https://github.com/PowerShell/PowerShell/blob/master/src/System.Management.Automation/engine/Modules/ModuleCmdletBase.cs#L6866). And it looks to me the registered `JobSourceAdapter` types never got unregistered when the module is removed ...

I don't think we should update the engine code to register 3rd party download protocol implementation during module importing. It should be more explicit and something done by the module itself.
I guess providing cmdlets for registration/unregistration also make sense, so a module author can also choose to do register in `.psm1` file, and unregister with the `OnRemove` property of `PSModuleInfo`.